lib/spack/spack/compiler.py

```diff
@@ -607,6 +607,10 @@ def _compiler_environment(self):
                 # See: https://github.com/spack/spack/issues/3153
                 if os.environ.get("CRAY_CPU_TARGET") == 'mic-knl':
                     spack.util.module_cmd.load_module('cce')
+
+                if not spack.util.module_cmd.module('avail', module):
```

Comment:
I'd be concerned that this potentially masks other issues. Maybe
```suggestion
                if not spack.util.module_cmd.module('avail', module).strip().endswith(module):
```
would be better to ensure that the module requested actually exists, and the returned string is not just an error?

Reply:
yes, the patch definitely needs improvement (wrt your comment in https://github.com/spack/spack/issues/18606#issuecomment-767610044). I assumed that at least the return code is checked for, but that is also not the case (should be `127` for a `command not found`).

But since the module matching can be fuzzy:

```console
$ module avail intel

----------------------- /data/tiziano/spack/share/spack/modules/linux-opensuse_leap15-sandybridge -----------------------
intel-parallel-studio-cluster.2020.2-intel-19.1.2.254-jzepir2

----------------------------------------------- /sw/modules/generic/cmpl ------------------------------------------------
intel/studio15/0.090 intel/studio16/3.210 intel/studio17/2.174 intel/studio17/4.196
$ module load intel
$ module list
Currently Loaded Modulefiles:
  1) intel/studio17/4.196
```

the suggested test is an improvement but may still not catch what is intended.

If we want to allow fuzzy matching, then I'd suggest:
```python
   if module not in spack.util.module_cmd.module('avail', module):
```

and if we do not:

```python
   if len(m for m in spack.util.module_cmd.module('avail', module).split() if module in m) != 1: 
```

or maybe more performant:

```python
modules = re.finditer(module, spack.util.module_cmd.module('avail', module))
try:
    next(modules)
except:
    raise ModuleNotFoundError(...)

try:
    next(modules)
except:
    pass
else:
    raise MoreThanOneModuleFound(...)
```

Reply:
Ok, I used a different approach now: load the module and then verify it has been loaded by checking the `LOADEDMODULES` env var. This way we don't depend on any output or any return values which may or may not be set.